client/clientservice/src/client_service.cpp

```diff
@@ -15,22 +15,66 @@
 
 namespace concord::client::clientservice {
 
-void ClientService::start(const std::string& addr, uint64_t max_receive_msg_size) {
+void ClientService::start(const std::string& addr, int num_async_threads, uint64_t max_receive_msg_size) {
   grpc::EnableDefaultHealthCheckService(true);
 
   grpc::ServerBuilder builder;
   builder.AddListeningPort(addr, grpc::InsecureServerCredentials());
   builder.SetMaxReceiveMessageSize(max_receive_msg_size);
-  builder.RegisterService(request_service_.get());
+
+  // Register synchronous services
   builder.RegisterService(event_service_.get());
 
+  // Register asynchronous services
+  builder.RegisterService(&request_service_);
+
+  for (int i = 0; i < num_async_threads; ++i) {
+    cqs_.emplace_back(builder.AddCompletionQueue());
+  }
+
   auto clientservice_server = std::unique_ptr<grpc::Server>(builder.BuildAndStart());
 
+  // From the "C++ Performance Notes" in the gRPC documentation:
+  // "Right now, the best performance trade-off is having numcpu's threads and one completion queue per thread."
+  for (int i = 0; i < num_async_threads; ++i) {
+    server_threads_.emplace_back(std::thread([this, i] { this->handleRpcs(i); }));
+  }
+
   auto health = clientservice_server->GetHealthCheckService();
   health->SetServingStatus(kRequestService, true);
   health->SetServingStatus(kEventService, true);
 
   clientservice_server->Wait();
+
+  LOG_INFO(logger_, "Wait for async gRPC threads to finish");
+  for (auto& t : server_threads_) {
+    t.join();
+  }
+}
+
+void ClientService::handleRpcs(int thread_idx) {
+  // Note: Memory is freed in `proceed`
+  new requestservice::RequestServiceCallData(&request_service_, cqs_[thread_idx].get(), client_);
```

Comment:
How is the `RequestServiceCallData` object itself freed? It seems we never assign the pointer to the allocated memory. Do I miss something?

Reply:
I saw the `delete this` in proceed(). I am wondering if this pattern is exception-safe? The way I understand it is that the pointer is passed around until we reach the end state, but I was thinking if we never do, due to an exception.

In my mind, it is confusing and risky, but I might be missing something and I haven't used a similar one before. What if we use a shared_ptr to pass it around?

Reply:
One more case I can think of is when `handleRpcs()` returns early - then there will be a memory leak, if I am not mistaken.

Reply:
I'm following the gRPC example applications and have looked into others as well. This seems to be the pattern for async services. I will add a try-catch block and schedule an alarm to reach the finish state.
With shared pointers, we could have the clientservice own the calldata for all ongoing requests. Meaning, we maintain a list/map of pointers that we have to protect from concurrent access or we make them thread-local 🤔